controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -141,6 +141,20 @@ func (r *BareMetalHostReconciler) Reconcile(ctx context.Context, request ctrl.Re
 		}
 	}
 
+	// If DataImage exists, add its ownerReference
+	dataImage := &metal3api.DataImage{}
+	err = r.Get(ctx, request.NamespacedName, dataImage)
+	if !(err != nil || ownerReferenceExists(host, dataImage)) {
+		if err := controllerutil.SetControllerReference(host, dataImage, r.Scheme()); err != nil {
```

Comment:
We should never set a Controller reference. Use `SetOwnerReference` always and everywhere.

Reply:
Looking at go docs, `SetControllerReference` triggers reconciliation for owner object when controlled object changes - is that the only issue or are there some other side effects as well which are not desirable in this case ?

Reply:
Owner reference is enough to trigger reconciliation iff you have `Owns()` (e.g. https://github.com/hroyrh/baremetal-operator/blob/0efbb139a9dfe3b5011780bd725968f7b76c91ff/controllers/metal3.io/baremetalhost_controller.go#L2294) in the controller setup.
A Controller reference is just an owner reference with the `controller` flag set. This makes the reference mutually exclusive. It exists to [solve a specific use case for ReplicaSet](https://github.com/kubernetes/design-proposals-archive/blob/acc25e14ca83dfda4f66d8cb1f1b491f26e78ffe/api-machinery/controller-ref.md) that ~nobody else has.